htsget_server/variants.py

```diff
@@ -0,0 +1,326 @@
+    return ((b * (b + 1)) / 2) + a
+
+
+def seq_match(a, b):
+    return len(set(expand_iupac(a)).intersection(set(expand_iupac(b)))) > 0
+
+
+def expand_iupac(base_str):
+    result = []
+    ambig_match = re.match(r"(^.*?)([RYSWKMBDHVN])(.*$)", base_str)
+    if ambig_match is None:
+        return [base_str]
+    # while there is an ambiguity char in group 2, append the expanded seqs to result, then run expand_iupac on each of those results
+    # ooh, apparently python 3.10 has match/case:
+    match ambig_match.group(2):
+        case 'R':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+        case 'Y':
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'S':
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+        case 'W':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'K':
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'M':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+        case 'B':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'D':
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'H':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+        case 'V':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+        case 'N':
+            result.append(f"{ambig_match.group(1)}A{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}C{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}G{ambig_match.group(3)}")
+            result.append(f"{ambig_match.group(1)}T{ambig_match.group(3)}")
+    final = []
+    while len(result) > 0:
+        final.extend(expand_iupac(result.pop(0)))
```

Comment:
This function appears to recurse once per ambiguous nucleotide -- is there any risk of stack overflows while reading areas with lots of ambiguity?

Reply:
In theory, but in the case of VCF files and searches, the files wouldn't have that level of ambiguity. They wouldn't have successfully aligned to the reference at all in that case.